File: client/app/server/routers/team.ts
import { protectedProcedure } from "~/server/trpc";
import { z } from "zod";
import { teamsApiFactory } from "~/server/api/clients/server";
import { enrichMembersWithAuthDetails } from "~/server/services/users";

export const teamRouter = {
  createTeam: protectedProcedure
    .input(
      z.object({
        organizationId: z.number(),
        name: z.string(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .createTeam(userId, input.organizationId, {
          slug: input.name,
        })
        .then((res) => res.data);
    }),
  getUserTeams: protectedProcedure
    .input(
      z.object({
        organizationId: z.number(),
      }),
    )
    .query(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .getUserTeams(userId, input.organizationId)
        .then((res) => res.data);
    }),
  getTeamMembers: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
      }),
    )
    .query(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      const members = await teamsApiFactory
        .getTeamMembers(userId, input.teamId)
        .then((res) => res.data);

      return await enrichMembersWithAuthDetails(members);
    }),
  joinTeam: protectedProcedure
    .input(
      z.object({
        organizationId: z.number(),
        slug: z.string(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .joinTeam(userId, input.organizationId, {
          slug: input.slug,
        })
        .then((res) => res.data);
    }),
  addTeamMember: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        userId: z.number(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const callerId = ctx.user?.id;
      return await teamsApiFactory
        .addTeamMember(callerId, input.teamId, { userId: input.userId })
        .then((res) => res.data);
    }),
  removeTeamMember: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        userId: z.number(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const callerId = ctx.user?.id;
      return await teamsApiFactory
        .removeTeamMember(callerId, input.teamId, { userId: input.userId })
        .then((res) => res.data);
    }),
  getTeamRepositories: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
      }),
    )
    .query(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .getTeamRepositories(userId, input.teamId)
        .then((res) => res.data);
    }),
  assignRepoToTeam: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        githubRepoId: z.number(),
        repoName: z.string(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .assignRepoToTeam(userId, input.teamId, {
          githubRepoId: input.githubRepoId,
          repoName: input.repoName,
        })
        .then((res) => res.data);
    }),
  unassignRepoFromTeam: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        githubRepoId: z.number(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .unassignRepoFromTeam(userId, input.teamId, input.githubRepoId)
        .then((res) => res.data);
    }),
  getTeamClusters: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
      }),
    )
    .query(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .getTeamClusters(userId, input.teamId)
        .then((res) => res.data);
    }),
  assignClusterToTeam: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        clusterId: z.number(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .assignClusterToTeam(userId, input.teamId, input.clusterId)
        .then((res) => res.data);
    }),
  unassignClusterFromTeam: protectedProcedure
    .input(
      z.object({
        teamId: z.number(),
        clusterId: z.number(),
      }),
    )
    .mutation(async ({ input, ctx }) => {
      const userId = ctx.user?.id;
      return await teamsApiFactory
        .unassignClusterFromTeam(userId, input.teamId, input.clusterId)
        .then((res) => res.data);
    }),
};

File: server/internal/api/application/team.go
package application

import (
	"context"
	"errors"

	interfaces "starliner.app/internal/api/domain/repository/interface"
	"starliner.app/internal/api/domain/service"
	"starliner.app/internal/api/domain/value"
	coreService "starliner.app/internal/core/domain/service"
)

type TeamApplication struct {
	teamRepository       interfaces.TeamRepository
	clusterRepository    interfaces.ClusterRepository
	githubAppRepository  interfaces.GithubAppRepository
	organizationService  *service.OrganizationService
	normalizationService *coreService.NormalizerService
	teamService          *service.TeamService
}

func NewTeamApplication(
	teamRepository interfaces.TeamRepository,
	clusterRepository interfaces.ClusterRepository,
	githubAppRepository interfaces.GithubAppRepository,
	organizationService *service.OrganizationService,
	normalizationService *coreService.NormalizerService,
	teamService *service.TeamService,
) *TeamApplication {
	return &TeamApplication{
		teamRepository:       teamRepository,
		clusterRepository:    clusterRepository,
		githubAppRepository:  githubAppRepository,
		organizationService:  organizationService,
		normalizationService: normalizationService,
		teamService:          teamService,
	}
}

func (ta *TeamApplication) CreateTeam(ctx context.Context, slug string, organizationId int64, userId int64) (*value.Team, error) {
	err := ta.organizationService.ValidateUserOrgOwner(ctx, organizationId, userId)
	if err != nil {
		return nil, err
	}

	normalized, err := ta.normalizationService.FormatToDNS1123(slug)
	if err != nil {
		return nil, err
	}

	if normalized != slug {
		return nil, errors.New("invalid slug format")
	}

	team, err := ta.teamRepository.CreateTeam(
		ctx,
		slug,
		organizationId,
	)
	if err != nil {
		return nil, err
	}

	err = ta.teamRepository.AddTeamMember(ctx, team.Id, userId)
	if err != nil {
		return nil, err
	}

	return value.NewTeam(team), nil
}

func (ta *TeamApplication) GetUserTeams(ctx context.Context, organizationId int64, userId int64) ([]*value.Team, error) {
	err := ta.organizationService.ValidateUserInOrg(ctx, organizationId, userId)
	if err != nil {
		return nil, err
	}

	teams, err := ta.teamRepository.GetUserTeams(ctx, organizationId, userId)
	if err != nil {
		return nil, err
	}

	return value.NewTeams(teams), nil
}

func (ta *TeamApplication) GetTeamMembers(ctx context.Context, userId int64, teamId int64) ([]*value.User, error) {
	err := ta.teamService.ValidateUserInTeam(ctx, userId, teamId)
	if err != nil {
		return nil, err
	}

	members, err := ta.teamRepository.GetTeamMembers(ctx, teamId)
	if err != nil {
		return nil, err
	}

	return value.NewUsers(members), nil
}

func (ta *TeamApplication) AddTeamMember(ctx context.Context, userId int64, teamId int64, callerId int64) error {
	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, team.OrganizationId, callerId)
	if err != nil {
		return err
	}

	return ta.teamRepository.AddTeamMember(ctx, teamId, userId)
}

func (ta *TeamApplication) JoinTeam(ctx context.Context, organizationId int64, userId int64, teamSlug string) error {
	err := ta.organizationService.ValidateUserInOrg(ctx, organizationId, userId)
	if err != nil {
		return err
	}

	team, err := ta.teamRepository.GetTeamBySlug(ctx, teamSlug, organizationId)
	if err != nil {
		return err
	}

	return ta.teamRepository.AddTeamMember(ctx, team.Id, userId)
}

func (ta *TeamApplication) RemoveTeamMember(ctx context.Context, userId int64, teamId int64, callerId int64) error {
	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, team.OrganizationId, callerId)
	if err != nil {
		return err
	}

	// Only the org owner is allowed to manage team members and is part of every team.
	// This check enforces the owner does not remove himself from a team.
	if callerId == userId {
		return errors.New("organization owner cannot be removed from team")
	}

	err = ta.teamRepository.RemoveTeamMember(ctx, teamId, userId)
	if err != nil {
		return err
	}

	return ta.teamRepository.DeleteTeamIfEmpty(ctx, teamId)
}

func (ta *TeamApplication) AssignRepoToTeam(ctx context.Context, userId int64, teamId int64, githubRepoId int64, repoName string) error {
	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, team.OrganizationId, userId)
	if err != nil {
		return err
	}

	githubApp, err := ta.githubAppRepository.GetOrganizationGithubApp(ctx, team.OrganizationId)
	if err != nil {
		return err
	}

	return ta.teamRepository.AssignRepoToTeam(ctx, teamId, githubRepoId, repoName, githubApp.ID)
}

func (ta *TeamApplication) UnassignRepoFromTeam(ctx context.Context, userId int64, teamId int64, githubRepoId int64) error {
	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, team.OrganizationId, userId)
	if err != nil {
		return err
	}

	return ta.teamRepository.UnassignRepoFromTeam(ctx, teamId, githubRepoId)
}

func (ta *TeamApplication) GetTeamRepositories(ctx context.Context, userId int64, teamId int64) ([]*value.TeamRepo, error) {
	err := ta.teamService.ValidateUserInTeam(ctx, userId, teamId)
	if err != nil {
		return nil, err
	}
	repos, err := ta.teamRepository.GetTeamRepositories(ctx, teamId)
	if err != nil {
		return nil, err
	}

	return value.NewTeamRepos(repos), nil
}

func (ta *TeamApplication) GetTeamClusters(ctx context.Context, userId int64, teamId int64) ([]*value.TeamCluster, error) {
	err := ta.teamService.ValidateUserInTeam(ctx, userId, teamId)
	if err != nil {
		return nil, err
	}
	clusters, err := ta.teamRepository.GetTeamClusters(ctx, teamId)
	if err != nil {
		return nil, err
	}
	return value.NewTeamClusters(clusters), nil
}

func (ta *TeamApplication) AssignClusterToTeam(ctx context.Context, userId int64, teamId int64, clusterId int64) error {
	// TODO: reduce amount of queries
	cluster, err := ta.clusterRepository.GetCluster(ctx, clusterId)
	if err != nil {
		return err
	}

	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	if cluster.OrganizationId != team.OrganizationId {
		return errors.New("cluster and team belong to different organizations")
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, cluster.OrganizationId, userId)
	if err != nil {
		return err
	}

	return ta.teamRepository.AssignClusterToTeam(ctx, teamId, clusterId)
}

func (ta *TeamApplication) UnassignClusterFromTeam(ctx context.Context, userId int64, teamId int64, clusterId int64) error {
	// TODO: reduce amount of queries
	cluster, err := ta.clusterRepository.GetCluster(ctx, clusterId)
	if err != nil {
		return err
	}

	team, err := ta.teamRepository.GetTeamById(ctx, teamId)
	if err != nil {
		return err
	}

	if cluster.OrganizationId != team.OrganizationId {
		return errors.New("cluster and team belong to different organizations")
	}

	err = ta.organizationService.ValidateUserOrgOwner(ctx, cluster.OrganizationId, userId)
	if err != nil {
		return err
	}

	return ta.teamRepository.UnassignClusterFromTeam(ctx, teamId, clusterId)
}
